### PR TITLE
Switch Loki to StatefulSet

### LINKED ISSUE #570: Helm: Loki stuck in ContainerCreating stage due to PVC
**Describe the bug**
When using helm upgrade using PVC, Loki stays stuck as ContainerCreating stage due to `Unable to mount volumes for pod`.

**To Reproduce**
1. Helm install Loki
2. Helm upgrade Loki with any change to config or resources
3. run `kubectl get pods` and never get results

```sh
loki-5f9d8554f-wqgmk   1/1     Running             0          1h
loki-d694cfc6-ccfvf    0/1     ContainerCreating   0          1h
```

Describing the stuck pod shows

```
  Warning  FailedMount  3s (x45 over 100m)  kubelet, gke-crawlers-pool-1-2005c642-8ks9  Unable to mount volumes for pod "loki-d694cfc6-ccfvf_loki(e44ce3e3-7326-11e9-a5cf-42010a8e0212)": timeout expired waiting for volumes to attach or mount for pod "loki"/"loki-d694cfc6-ccfvf". list of unmounted volumes=[storage]. list of unattached volumes=[config storage loki-token-dpkvn]
```

Should we use StatefulSet instead of Deployment types since they're better at mounting and unmounting  and general management of PVCs?

**Expected behavior**

I expect the new release to just work

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: helm

### PR DESCRIPTION
**Pull request by steven-sheehy**
**What this PR does / why we need it**:
This switches the Loki `Deployment` to a `StatefulSet`, which is more appropriate when used with persistent volumes. If not using a PVC (e.g. in memory or cloud storage backends), then it will still use an emptyDir making it compatible with that approach.

- Removed non-StatefulSet fields `minReadySeconds` and `deploymentStrategy`
- Added StatefulSet fields `podManagementPolicy`, `serviceName` and `updateStrategy`
- Added a headless service as is required for a StatefulSet to have stable network identity

**Which issue(s) this PR fixes**:
Fixes #570

**Special notes for your reviewer**:

**WARNING!** This is a breaking change and will delete the old PVC and create a new PVC, losing all of the previous data in the process. This is unfortunate but it's better to do this now while it's in alpha. Users should backup their current PVC data and restore it manually after upgrading if that data is important.

**Checklist**
- [x] Documentation added
- [ ] Tests updated

### CODE REVIEW

**Comment by steven-sheehy** [comment]
@sboschman Would appreciate a review and some manual testing.

**Comment by sboschman** [comment]
LGTM

Upgraded my setup (1x distributor, 3x ingester, 3x querier and 1x tablemanager) with the statefulset from this PR and seems to work allright.

**Comment by steven-sheehy** [comment]
Thanks. @Kuqd This is ready for review.

File: production/helm/loki/templates/service-headless.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+apiVersion: v1
```
Comment (rdxmb) [inline production/helm/loki/templates/service-headless.yaml]: what is the headless service for in this repository? Are there any pods communicating over this service?

File: production/helm/loki/templates/statefulset.yaml [hunk 2]
```diff
@@ -83,11 +81,25 @@ spec:
...
-      {{- if .Values.persistence.enabled }}
-          persistentVolumeClaim:
-            claimName: {{ .Values.persistence.existingClaim | default (include "loki.fullname" .) }}
-      {{- else }}
           emptyDir: {}
-      {{- end }}
+  {{- else if .Values.persistence.existingClaim }}
+        - name: storage
+          persistentVolumeClaim:
+            claimName: {{ .Values.persistence.existingClaim }}
```
Comment (rdxmb) [inline production/helm/loki/templates/statefulset.yaml]: does this work with more than one replica - so they all use the same existingClaim?

Comment (steven-sheehy) [reply production/helm/loki/templates/statefulset.yaml]: Yes, if that persistent volume bound to the PVC supports `ReadWriteMany` access mode like NFS. It us up to the user to provision the PV and PVC.

File: production/helm/loki/templates/service-headless.yaml (hunk 1 above)
Comment (steven-sheehy) [reply production/helm/loki/templates/service-headless.yaml]: As mentioned in the description, it's required for a `StatefulSet`. Also, it provides a stable hostname which allows consul to be used as a ring store.

File: production/helm/loki/templates/statefulset.yaml (hunk 2 above)
Comment (rdxmb) [reply production/helm/loki/templates/statefulset.yaml]: Yes, but I wonder if the application (loki) does support that ...

Comment (sboschman) [reply production/helm/loki/templates/statefulset.yaml]: The chunk filenames are uuid's, so writes in the same dir do not collide. At least haven't seen any issues yet with 5k lines per sec in my setup.

Comment (rdxmb) [reply production/helm/loki/templates/statefulset.yaml]: ok

**Comment by Kuqd** [comment]
Can you rebase, I'll take care of this today !

**Comment by steven-sheehy** [comment]
@Kuqd Done

**Pull request merged**